server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/PartitionCache.java

```diff
@@ -0,0 +1,710 @@
+  /** TConsensusGroupId -> TRegionReplicaSet */
+  private final Map<TConsensusGroupId, TRegionReplicaSet> groupIdToReplicaSetMap =
+      new ConcurrentHashMap<>();
+
+  /** The lock of cache */
+  private final ReentrantReadWriteLock storageGroupCacheLock = new ReentrantReadWriteLock();
+
+  private final ReentrantReadWriteLock schemaPartitionCacheLock = new ReentrantReadWriteLock();
+  private final ReentrantReadWriteLock dataPartitionCacheLock = new ReentrantReadWriteLock();
+
+  private final IClientManager<PartitionRegionId, ConfigNodeClient> configNodeClientManager =
+      new IClientManager.Factory<PartitionRegionId, ConfigNodeClient>()
+          .createClientManager(new DataNodeClientPoolFactory.ConfigNodeClientPoolFactory());
+
+  public PartitionCache() {
+    this.schemaPartitionCache = Caffeine.newBuilder().maximumSize(cacheSize).build();
+    this.dataPartitionCache = Caffeine.newBuilder().maximumSize(cacheSize).build();
+    this.partitionExecutor =
+        SeriesPartitionExecutor.getSeriesPartitionExecutor(
+            this.seriesSlotExecutorName, this.seriesPartitionSlotNum);
+  }
+
+  // region storage group cache
+
+  /** get storage group of device, return null if not cache */
+  public String getStorageGroup(String devicePath) {
+    synchronized (storageGroupCache) {
+      for (String storageGroup : storageGroupCache) {
+        if (PathUtils.isStartWith(devicePath, storageGroup)) {
+          return storageGroup;
+        }
+      }
+    }
+    return null;
+  }
+
+  /** get device to storage group map */
+  public boolean getDeviceToStorageGroupMap(
+      List<String> devicePaths,
+      Map<String, String> deviceToStorageGroupMap,
+      List<String> failedDevices,
+      boolean failFast) {
+    storageGroupCacheLock.readLock().lock();
+    try {
+      boolean result = true;
+      for (String devicePath : devicePaths) {
+        String storageGroup = getStorageGroup(devicePath);
+        if (null == storageGroup) {
+          logger.debug(
+              "[{} Cache] miss when search device {}", STORAGE_GROUP_CACHE_NAME, devicePath);
+          result = false;
+          if (failFast) {
+            break;
+          } else {
+            failedDevices.add(devicePath);
+          }
+        } else {
+          deviceToStorageGroupMap.put(devicePath, storageGroup);
+        }
+      }
+      if (!result) {
+        deviceToStorageGroupMap.clear();
+      }
+      logger.debug("[{} Cache] hit when search device {}", STORAGE_GROUP_CACHE_NAME, devicePaths);
+      CacheMetricsRecorder.record(result, STORAGE_GROUP_CACHE_NAME);
+      return result;
+    } finally {
+      storageGroupCacheLock.readLock().unlock();
+    }
+  }
+
+  /** get storage group to device map with failed devices */
+  public boolean getStorageGroupToDeviceMap(
+      List<String> devicePaths,
+      Map<String, List<String>> storageGroupToDeviceMap,
+      List<String> failedDevices,
+      boolean failFast) {
```

Comment:
We'd better add function annotation to explain what these input parameters for and what this interface do since this is a **public** interface~

Reply:
Get it ~